go/vt/schemamanager/tablet_executor.go

```diff
@@ -190,6 +195,76 @@ func (exec *TabletExecutor) isOnlineSchemaDDL(stmt sqlparser.Statement) (isOnlin
 	return false
 }
 
+func validateThrottleParams(alterMigrationType sqlparser.AlterMigrationType, expireString string, ratioLiteral *sqlparser.Literal) (duration time.Duration, ratio float64, err error) {
+	switch alterMigrationType {
+	case sqlparser.UnthrottleMigrationType,
+		sqlparser.UnthrottleAllMigrationType:
+		// Unthrottling is like throttling with duration=0
+		duration = 0
+	default:
+		duration = throttle.DefaultAppThrottleDuration
+		if expireString != "" {
+			duration, err = time.ParseDuration(expireString)
+			if err != nil || duration < 0 {
+				return duration, ratio, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid EXPIRE value: %s. Try '120s', '30m', '1h', etc. Allowed units are (s)ec, (m)in, (h)hour", expireString)
+			}
+		}
+	}
+	ratio = 1.0
+	if ratioLiteral != nil {
+		ratio, err = strconv.ParseFloat(ratioLiteral.Val, 64)
+		if err != nil || ratio < 0 || ratio > 1 {
+			return duration, ratio, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid RATIO value: %s. Try any decimal number between '0.0' (no throttle) and `1.0` (fully throttled)", ratioLiteral.Val)
+		}
+	}
+	return duration, ratio, nil
+}
+
+func (exec *TabletExecutor) executeAlterMigrationThrottle(ctx context.Context, alterMigration *sqlparser.AlterMigration) (err error) {
+	duration, ratio, err := validateThrottleParams(alterMigration.Type, alterMigration.Expire, alterMigration.Ratio)
+	if err != nil {
+		return err
+	}
+	expireAt := time.Now().Add(duration)
+	appName := alterMigration.UUID
+	if appName == "" {
+		appName = throttlerapp.OnlineDDLName.String()
+	}
+	throttledAppRule := &topodatapb.ThrottledAppRule{
+		Name:      appName,
+		ExpiresAt: protoutil.TimeToProto(expireAt),
+		Ratio:     ratio,
+	}
+
+	req := &vtctldatapb.UpdateThrottlerConfigRequest{
+		Keyspace:     exec.keyspace,
+		ThrottledApp: throttledAppRule,
+	}
+
+	update := func(throttlerConfig *topodatapb.ThrottlerConfig) *topodatapb.ThrottlerConfig {
+		if throttlerConfig == nil {
+			throttlerConfig = &topodatapb.ThrottlerConfig{}
+		}
+		if throttlerConfig.ThrottledApps == nil {
+			throttlerConfig.ThrottledApps = make(map[string]*topodatapb.ThrottledAppRule)
+		}
+		throttlerConfig.ThrottledApps[req.ThrottledApp.Name] = req.ThrottledApp
+		return throttlerConfig
+	}
+	// We have already locked the keyspace
+	ki, err := exec.ts.GetKeyspace(ctx, req.Keyspace)
+	if err != nil {
+		return err
+	}
```

Comment:
Where do we do that in this call path? We lock it in vtgate? In this PR it looks like the vtgate parses the incoming query in the query service and then calls this function.

Reply:
If we do lock it in vtgate, then that lock info should be attached to the context we get and we can use `CheckKeyspaceLocked` to be sure we have locked it:
```
if err = topo.CheckKeyspaceLocked(ctx, req.Keyspace); err != nil {
	return err
}
```

That would protect us in the case that somehow the locks lease expired or someone executed this directly against a tablet using one of the `Execute` RPCs.

Reply:
Ah, we're already doing just that here: https://github.com/planetscale/vitess/blob/0d8ca1be54ec50feda41b75e750619df15c5b747/go/vt/topo/srv_keyspace.go#L397-L401

So the safety aspect is already addressed. From there it's only my curiosity. 🙂

Reply:
@mattlord it's locked here:

https://github.com/vitessio/vitess/blob/ab6c7af2df4763e060bbc979d1ba5a2f7ee55e9b/go/vt/schemamanager/tablet_executor.go#L302